tests/integration_test/service_control_apikey_restriction_test/service_control_apikey_restriction_test.go

```diff
@@ -173,6 +176,62 @@ func TestServiceControlAPIKeyIpRestriction(t *testing.T) {
 	}
 }
 
+func TestExtractClientIPFromForwardedHeader(t *testing.T) {
+	serviceName := "test-echo"
+	configID := "test-config-id"
+	args := []string{
+		"--service=" + serviceName,
+		"--service_config_id=" + configID,
+
+		"--client_ip_from_forwarded_header",
+	}
+
+	s := env.NewTestEnv(platform.TestExtractClientIPFromForwardedHeader, platform.EchoSidecar)
+
+	defer s.TearDown(t)
+	if err := s.Setup(args); err != nil {
+		t.Fatalf("failed to setup test env, %v", err)
+	}
+
+	testData := []testDataStruct{
+		{
+			desc:     "Without forwarded header, client IP from tcp connection.",
+			url:      fmt.Sprintf("http://%v:%v%v%v", platform.GetLoopbackAddress(), s.Ports().ListenerPort, "/echo", "?key=api-key1"),
+			message:  "hello",
+			wantResp: `{"message":"hello"}`,
+			wantScRequest: &utils.ExpectedCheck{
+				Version:         utils.ESPv2Version(),
+				ServiceName:     "echo-api.endpoints.cloudesf-testing.cloud.goog",
+				ServiceConfigID: "test-config-id",
+				ConsumerID:      "api_key:api-key1",
+				OperationName:   "1.echo_api_endpoints_cloudesf_testing_cloud_goog.Echo",
+				ApiKey:          "api-key1",
+				CallerIp:        platform.GetLoopbackAddress(),
+			},
+		},
+		{
+			desc:            "client ip from forwarded header",
+			url:             fmt.Sprintf("http://%v:%v%v%v", platform.GetLoopbackAddress(), s.Ports().ListenerPort, "/echo", "?key=api-key3"),
+			message:         "hello",
+			forwardedHeader: "for=1.2.3.4",
+			wantResp:        `{"message":"hello"}`,
+			wantScRequest: &utils.ExpectedCheck{
+				Version:         utils.ESPv2Version(),
+				ServiceName:     "echo-api.endpoints.cloudesf-testing.cloud.goog",
+				ServiceConfigID: "test-config-id",
+				ConsumerID:      "api_key:api-key3",
+				OperationName:   "1.echo_api_endpoints_cloudesf_testing_cloud_goog.Echo",
+				ApiKey:          "api-key3",
+				CallerIp:        "1.2.3.4",
+			},
+		},
```

Comment:
May be worth adding a test around using forwarded even if x-forwarded-for is also specified?

Reply:
Done